tests/test_pre_shape.py

```diff
@@ -0,0 +1,239 @@
+        expected = Matrices.transpose(result)
+
+        self.assertAllClose(result, expected)
+
+    @geomstats.tests.np_and_pytorch_only
+    def test_horizontal_and_is_horizontal(self):
+        w = gs.random.rand(self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform()
+        tan = self.space.to_tangent(w, point)
+        horizontal = self.space.horizontal_projection(tan, point)
+
+        horizontal = gs.stack([horizontal, w])
+        result = self.space.is_tangent(horizontal, point)
+        expected = gs.array([True, False])
+
+        self.assertAllClose(result, expected)
+
+    def test_realign(self):
+        point, base_point = self.space.random_uniform(2)
+        aligned = self.space.realign(point, base_point)
+        alignment = gs.matmul(aligned, Matrices.transpose(base_point))
+        result = Matrices.is_symmetric(alignment)
+        self.assertTrue(result)
+
+    def test_realign_vectorization(self):
+        base_point = self.space.random_uniform()
+        point = self.space.random_uniform(2)
+        aligned = self.space.realign(point, base_point)
+        alignment = gs.matmul(aligned, Matrices.transpose(base_point))
+        result = Matrices.is_symmetric(alignment)
+        self.assertTrue(gs.all(result))
+
+        base_point = self.space.random_uniform(2)
+        point = self.space.random_uniform()
+        aligned = self.space.realign(point, base_point)
+        alignment = gs.matmul(aligned, Matrices.transpose(base_point))
+        result = Matrices.is_symmetric(alignment)
+        self.assertTrue(gs.all(result))
+
+    def test_inner_product(self):
+        w = gs.random.rand(self.n_samples, self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform()
+        tan = self.space.to_tangent(w, point)
+        inner = self.space.metric.inner_product(tan, tan, point)
+        self.assertAllClose(inner.shape, (self.n_samples,))
+
+    def test_exp_and_belongs(self):
+        w = gs.random.rand(self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform()
+        tan = self.space.to_tangent(w, point)
+        exp = self.space.metric.exp(tan, point)
+        result = self.space.belongs(exp)
+        self.assertTrue(result)
+
+        exp = self.space.metric.exp(gs.zeros_like(point), point)
+        result = gs.isclose(point, exp)
+        self.assertTrue(gs.all(result))
+
+    def test_exp_and_belongs_vectorization(self):
+        w = gs.random.rand(self.n_samples, self.m_ambient, self.k_landmarks)
+        point = self.space.random_uniform(self.n_samples)
+        tan = self.space.to_tangent(w, point)
+        exp = self.space.metric.exp(tan, point)
+        result = self.space.belongs(exp)
+        self.assertTrue(gs.all(result))
+
+        point = point[0]
+        tan = self.space.to_tangent(w, point)
+        exp = self.space.metric.exp(tan, point)
+        result = self.space.belongs(exp)
+        self.assertTrue(gs.all(result))
+
+    def test_log(self):
```

Comment:
This is `test_log_and_exp`

Reply:
Could we also add `test_exp_and_log`, i.e. first the Exponential and then the Logarithm?